src/effects/RealtimeEffectState.cpp

```diff
@@ -11,8 +11,126 @@
 #include "RealtimeEffectState.h"
 
 #include "EffectInterface.h"
+#include "MessageBuffer.h"
 #include "PluginManager.h"
 
+//! Mediator of two-way inter-thread communication of changes of settings
+class RealtimeEffectState::AccessState : public NonInterferingBase {
+public:
+   AccessState(EffectProcessor &effect, EffectSettings &settings)
+      : mEffect{effect}
+      , mSettings{settings}
+   {
+   }
+
+   void Initialize(const EffectSettings &settings) {
+      // Initialize each message buffer with two copies of settings
+      mChannelToMain.Write(ToMainSlot{settings});
+      mChannelToMain.Write(ToMainSlot{settings});
+      mChannelFromMain.Write(FromMainSlot{settings});
+      mChannelFromMain.Write(FromMainSlot{settings});
+   }
+
+   const EffectSettings &MainRead() {
+      // Main thread clones the object in the std::any, then gives a reference
+      mChannelToMain.Read<ToMainSlot::Reader>(mMainThreadCache);
+      return mMainThreadCache;
+   }
+   void MainWrite(EffectSettings &&settings) {
+      // Main thread may simply swap new content into place
+      mChannelFromMain.Write(std::move(settings));
+   }
+
+   struct EffectAndSettings{
+      EffectProcessor &effect; const EffectSettings &settings; };
+   void WorkerRead() {
+      // Worker thread avoids memory allocation.  It copies the contents of any
+      mChannelFromMain.Read<FromMainSlot::Reader>(mEffect, mSettings);
+   }
+   void WorkerWrite() {
+      // Worker thread avoids memory allocation.  It copies the contents of any
+      mChannelToMain.Write(EffectAndSettings{mEffect, mSettings});
+   }
+
+private:
+   struct ToMainSlot {
+      // For initialization of the channel
+      ToMainSlot() = default;
+      explicit ToMainSlot(const EffectSettings &settings)
+         // Copy std::any
+         : mSettings{ settings }
+      {}
+      ToMainSlot &operator=(ToMainSlot &&) = default;
+
+      // Worker thread writes the slot
+      ToMainSlot& operator=(EffectAndSettings &&arg) {
+         // This happens during MessageBuffer's busying of the slot
+         arg.effect.CopySettingsContents(arg.settings, mSettings);
```

Comment:
I'm still not sure that what `EffectSettings` object represents here. Is it supposed to hold values for every parameter that could be changed? If so, isn't it very expensive then?

Reply:
That is the idea.

For each of our built-in effects, that will only be a few numbers and booleans, and `CopySettingsContents` would be overriden just to copy them, rather than copying the std::any which implies memory allocation.